lib/flatpak-installation.c

```diff
@@ -1112,170 +1112,254 @@ get_metadata_progress (guint metadata_fetched,
+  gint fetched, total_fetches;
+
+  /* It would've been so much easier of OSTree just exposed all
+   * the progress report fields. All the following math is just
+   * the preparation for us to be able to do:
+   *
+   *   average_bytes_per_object = downloaded content bytes / downloaded content objects
+   */
+
+  /* All objects */
+  fetched = ostree_async_progress_get_uint (progress, "fetched");
+  total_fetches = fetched + ostree_async_progress_get_uint (progress, "outstanding-fetches");
+
+  /* Metadata objects */
+  fetched_metadata = ostree_async_progress_get_uint (progress, "metadata-fetched");
+  outstanding_metadata = ostree_async_progress_get_uint (progress, "outstanding-metadata-fetches");
+  total_metadata = fetched_metadata + outstanding_metadata;
+
+  /* Static deltas */
+  total_deltas = ostree_async_progress_get_uint (progress, "total-delta-parts")
+               + ostree_async_progress_get_uint (progress, "total-delta-fallbacks");
+  fetched_deltas_size = ostree_async_progress_get_uint64 (progress, "fetched-delta-part-size");
+
+  /* All the downloaded bytes, without the downloaded deltas */
+  metadata_bytes = ostree_async_progress_get_uint64 (progress, "last-metadata-bytes");
+  bytes_transferred = ostree_async_progress_get_uint64 (progress, "bytes-transferred");
+  bytes_transferred -= fetched_deltas_size;
+
+  /* While downloading metadata, store the currently downloaded size
+   * and don't report anything */
+  if (outstanding_metadata > 0 || metadata_bytes == 0)
+    {
+      /* Keep track of when OSTree started to download content objects */
+      ostree_async_progress_set_uint64 (progress, "last-metadata-bytes", bytes_transferred);
+      return;
+    }
+
+  /* Finally, the content-related stuff */
+  content_bytes = bytes_transferred - metadata_bytes;
+  total_content = total_fetches - total_metadata - total_deltas;
+  fetched_content = fetched - fetched_metadata;
+
+  /* Only report content progress when there was content */
+  if (total_content > 0 && fetched_content > 0)
+    {
+      gdouble average_bytes;
+
+      average_bytes = content_bytes / (gdouble) fetched_content;
+
+      if (fetched_content_bytes)
+        *fetched_content_bytes = (guint64) content_bytes;
+
+      if (total_content_bytes)
+        *total_content_bytes = (guint64) total_content * average_bytes;
+    }
+}
+
+static void
+format_downloaded_bytes (OstreeAsyncProgress  *progress,
+                         guint64               current_time,
+                         gchar               **out_transferred,
+                         gchar               **out_bytes_per_sec)
+{
+  guint64 bytes_transferred = ostree_async_progress_get_uint64 (progress, "bytes-transferred");
+  guint64 elapsed_time =
+    (current_time - ostree_async_progress_get_uint64 (progress, "start-time")) / G_USEC_PER_SEC;
+  guint64 bytes_sec = (elapsed_time > 0) ? bytes_transferred / elapsed_time : 0;
+  gchar *formatted_bytes_transferred =
```

Comment:
I think Flatpak uses just _char_, as _gchar_ is just the same thing but originally for getting special highlight IIRC...